### ISSUE TITLE
AzuraCast not always returning titles of tracks played by AutoDJ

### ISSUE THREAD

**Issue by ivellios1988** [AzuraCast/AzuraCast#4434]
### Installation Method

Docker Installation

### AzuraCast Release Channel

Stable Channel

### Current AzuraCast Version

v0.13.0 Stable

### What happened?

I care a lot about ID3 tags properly set for files in my AutoDJ. However, on many occasions, I observe that even when the title of the next track to be played is properly shown in my AzuraCast admin panel, when the track is finally played it shows 'Unknown' in the title bar.

All MP3s and most of FLAC and AAC files in my AutoDJ have ID3 tags containing Artist and Title, sometimes also Album, Cover, Year, etc. The only exceptions are .opus files which don't have tags, instead, AzuraCast seems to treat filename as title and usually puts it properly into the title bar.

To tag my files, I use the tag editor in AzuraCast's file manager.

Here I attach one of the problematic 'tagged but unknown' MP3s: https://drive.google.com/file/d/1uKmzYUIxPbdhIfY71ipagMOMhnkrO_IV/view?usp=sharing

### Relevant log output

```shell
2021/07/28 15:29:23 [lang:3] AzuraCast Raw Response: annotate:title="Conquest of paradise (Vangelis)",artist="5 covers mixed together",duration="365.00",song_id="4ab2c2ef9276b004f07249568f94e0e3",media_id="12",liq_amplify="0.00dB",playlist_id="3":/var/azuracast/stations/radio_paranormalium/media/muzyka/covery/5 covers mixed together - Conquest of paradise (Vangelis).mp3
2021/07/28 15:29:23 [decoder:3] Method "FFMPEG" accepted "/var/azuracast/stations/radio_paranormalium/media/muzyka/covery/5 covers mixed together - Conquest of paradise (Vangelis).mp3".
2021/07/28 15:29:23 [ffmpeg:3] [mp3 @ 0x7fa88402acc0] Estimating duration from bitrate, this may be inaccurate
2021/07/28 15:29:23 [lang:3] AzuraCast Feedback Response: OK
2021/07/28 15:29:46 [ffmpeg:3] [mp3 @ 0x7fa87c1ab500] Estimating duration from bitrate, this may be inaccurate
2021/07/28 15:29:46 [ffmpeg:3] [mp3 @ 0x7fa87c1ab500] Estimating duration from bitrate, this may be inaccurate

[now a jingle is prepared]

2021/07/28 15:29:46 [next_song:3] Prepared "/var/azuracast/stations/radio_paranormalium/media/jingle/jingle czat.mp3" (RID 3).
2021/07/28 15:29:46 [cross_38793:3] Analysis: -30.295588dB / -14.928928dB (2.99s / 3.01s)
2021/07/28 15:29:46 [crossfade:3] Simple transition: crossed, fade-in, fade-out.
2021/07/28 15:29:47 [lang:3] AzuraCast Feedback Response: OK
2021/07/28 15:29:59 [ffmpeg:3] [mp3 @ 0x7fa87c021400] Estimating duration from bitrate, this may be inaccurate
2021/07/28 15:29:59 [ffmpeg:3] [mp3 @ 0x7fa87c021400] Estimating duration from bitrate, this may be inaccurate

[and finally time to play the song!]

2021/07/28 15:29:59 [next_song:3] Prepared "/var/azuracast/stations/radio_paranormalium/media/muzyka/covery/5 covers mixed together - Conquest of paradise (Vangelis).mp3" (RID 5).
2021/07/28 15:29:59 [cross_38793:3] Analysis: -31.089749dB / -28.217561dB (2.96s / 2.96s)
2021/07/28 15:29:59 [crossfade:3] Simple transition: crossed, fade-in, fade-out.
2021/07/28 15:29:59 [lang:3] AzuraCast Feedback Response: OK
2021/07/28 15:31:32 [server:3] New client: 172.19.0.5.
2021/07/28 15:31:32 [server:3] Client 172.19.0.5 disconnected.
```

**Comment by ivellios1988** [comment]
I should add that, on many occasions, I observe that when the same file is played a number of times, one time the title is shown properly, other time it is 'untitled'. Many times I see that the file currently played file is 'untitled' and its title is displayed as the title of the 'playing next'.

**Comment by ivellios1988** [comment]
The issue still occurs after updating AzuraCast to v0.14.0 Stable

**Comment by ivellios1988** [comment]
Here's my Liquidsoap configuration, maybe this will help finding the culprit:

`# WARNING! This file is automatically generated by AzuraCast.
# Do not update it directly!


set("init.daemon", false)
set("init.daemon.pidfile.path","/var/azuracast/stations/radio_paranormalium/config/liquidsoap.pid")
set("log.stdout", true)
set("log.file", false)
set("server.telnet",true)
set("server.telnet.bind_addr","0.0.0.0")
set("server.telnet.port", 8004)
set("harbor.bind_addrs",["0.0.0.0"])

set("tag.encodings",["UTF-8","ISO-8859-1"])
set("encoder.encoder.export",["artist","title","album","song"])

setenv("TZ", "Europe/Warsaw")

azuracast_api_auth = ref "(PASSWORD)"
ignore(azuracast_api_auth)

autodj_is_loading = ref true
ignore(autodj_is_loading)

autodj_ping_attempts = ref 0
ignore(autodj_ping_attempts)

playlist_muzyka_dzien = playlist(id="playlist_muzyka_dzien",mode="randomize",reload_mode="watch",conservative=true,default_duration=10.,length=20.,"/var/azuracast/stations/radio_paranormalium/playlists/playlist_muzyka_dzien.m3u")
playlist_muzyka_dzien = audio_to_stereo(id="stereo_playlist_muzyka_dzien", playlist_muzyka_dzien)
playlist_muzyka_dzien = cue_cut(id="cue_playlist_muzyka_dzien", playlist_muzyka_dzien)

playlist_muzyka_noc = playlist(id="playlist_muzyka_noc",mode="randomize",reload_mode="watch",conservative=true,default_duration=10.,length=20.,"/var/azuracast/stations/radio_paranormalium/playlists/playlist_muzyka_noc.m3u")
playlist_muzyka_noc = audio_to_stereo(id="stereo_playlist_muzyka_noc", playlist_muzyka_noc)
playlist_muzyka_noc = cue_cut(id="cue_playlist_muzyka_noc", playlist_muzyka_noc)

playlist_audycje_wszystkie = playlist(id="playlist_audycje_wszystkie",mode="randomize",reload_mode="watch",conservative=true,default_duration=10.,length=20.,"/var/azuracast/stations/radio_paranormalium/playlists/playlist_audycje_wszystkie.m3u")
playlist_audycje_wszystkie = audio_to_stereo(id="stereo_playlist_audycje_wszystkie", playlist_audycje_wszystkie)
playlist_audycje_wszystkie = cue_cut(id="cue_playlist_audycje_wszystkie", playlist_audycje_wszystkie)

playlist_jingle = playlist(id="playlist_jingle",mode="randomize",reload_mode="watch",conservative=true,default_duration=10.,length=20.,"/var/azuracast/stations/radio_paranormalium/playlists/playlist_jingle.m3u")
playlist_jingle = audio_to_stereo(id="stereo_playlist_jingle", playlist_jingle)
playlist_jingle = cue_cut(id="cue_playlist_jingle", playlist_jingle)

playlist_audycje_najnowsze = playlist(id="playlist_audycje_najnowsze",mode="randomize",reload_mode="watch",conservative=true,default_duration=10.,length=20.,"/var/azuracast/stations/radio_paranormalium/playlists/playlist_audycje_najnowsze.m3u")
playlist_audycje_najnowsze = audio_to_stereo(id="stereo_playlist_audycje_najnowsze", playlist_audycje_najnowsze)
playlist_audycje_najnowsze = cue_cut(id="cue_playlist_audycje_najnowsze", playlist_audycje_najnowsze)

# Standard Playlists
radio = random(id="standard_playlists", weights=[], [])

# Standard Schedule Switches

radio = switch(id="schedule_switch", track_sensitive=true, [ ({ 7h0m-22h0m }, playlist_muzyka_dzien), ({ (22h0m-23h59m59s) or (00h00m-7h0m) }, playlist_muzyka_noc), ({true}, radio) ])

# Once per x Songs Playlists
radio = rotate(weights=[1,5], [playlist_audycje_wszystkie, radio])
radio = rotate(weights=[1,3], [playlist_jingle, radio])
radio = rotate(weights=[1,7], [playlist_audycje_najnowsze, radio])

# AutoDJ Next Song Script
def autodj_next_song() =
    uri = list.hd(get_process_lines(env=[("API_AUTH", !azuracast_api_auth)], 'curl -s --request POST --url http://web/api/internal/1/nextsong --form api_auth="$API_AUTH"'), default="")
    log("AzuraCast Raw Response: #{uri}")

    if uri == "" or string.match(pattern="Error", uri) then
        []
    else
        r = request.create(uri)
        if request.resolve(r) then
            [r]
        else
            []
       end
    end
end

# Delayed ping for AutoDJ Next Song
def wait_for_next_song(autodj)
    autodj_ping_attempts := !autodj_ping_attempts + 1
    delay = ref 0.5

    if source.is_ready(!autodj) then
        log("AutoDJ is ready!")

        autodj_is_loading := false
        delay := -1.0
    elsif !autodj_ping_attempts > 200 then
        log("AutoDJ could not be initialized within the specified timeout.")

        autodj_is_loading := false
        delay := -1.0
    end

    !delay
end

dynamic = request.dynamic.list(id="next_song", timeout=20., retry_delay=2., autodj_next_song)
dynamic = audio_to_stereo(id="stereo_next_song", dynamic)
dynamic = cue_cut(id="cue_next_song", dynamic)

dynamic_startup = fallback(
    id = "dynamic_startup",
    track_sensitive = false,
    [
        dynamic,
        at(
            fun()-> !autodj_is_loading,
            blank(id = "autodj_startup_blank", duration = 120.)
        )
    ]
)
radio = fallback(id="autodj_fallback", track_sensitive = true, [dynamic_startup, radio])

ref_dynamic = ref dynamic;
add_timeout(0.25,fun()->wait_for_next_song(ref_dynamic))

requests = request.queue(id="requests")
requests = audio_to_stereo(id="stereo_requests", requests)
requests = cue_cut(id="cue_requests", requests)

radio = fallback(id="requests_fallback", track_sensitive = true, [requests, radio])
add_skip_command(radio)

radio = crossfade(smart=false, duration=3.00,fade_out=2.00,fade_in=2.00,radio)

# DJ Authentication
live_enabled = ref false
last_authenticated_dj = ref ""
live_dj = ref ""

def dj_auth(auth_user,auth_pw) =
    user = ref ""
    password = ref ""

    if (auth_user == "source" or auth_user == "") and (string.match(pattern="(:|,)+", auth_pw)) then
        auth_string = string.split(separator="(:|,)", auth_pw)

        user := list.nth(default="", auth_string, 0)
        password := list.nth(default="", auth_string, 2)
    else
        user := auth_user
        password := auth_pw
    end

    log("Authenticating DJ: #{!user}")

    ret = list.hd(get_process_lines(env=[("DJ_USER", !user), ("DJ_PASSWORD", !password), ("API_AUTH", !azuracast_api_auth)], 'curl -s --request POST --url http://web/api/internal/1/auth --form dj-user="$DJ_USER" --form dj-password="$DJ_PASSWORD" --form api_auth="$API_AUTH"'), default="")
    log("AzuraCast DJ Auth Response: #{ret}")

    authed = bool_of_string(ret)
    if (authed) then
        last_authenticated_dj := !user
    end

    authed
end

def live_connected(header) =
    dj = !last_authenticated_dj
    log("DJ Source connected! Last authenticated DJ: #{dj} - #{header}")

    live_enabled := true
    live_dj := dj

    ret = list.hd(get_process_lines(env=[("DJ_USER", dj), ("API_AUTH", !azuracast_api_auth)], 'curl -s --request POST --url http://web/api/internal/1/djon --form dj-user="$DJ_USER" --form api_auth="$API_AUTH"'), default="")
    log("AzuraCast Live Connected Response: #{ret}")
end

def live_disconnected() =
    dj = !live_dj

    log("DJ Source disconnected! Current live DJ: #{dj}")

    ret = list.hd(get_process_lines(env=[("DJ_USER", dj), ("API_AUTH", !azuracast_api_auth)], 'curl -s --request POST --url http://web/api/internal/1/djoff --form dj-user="$DJ_USER" --form api_auth="$API_AUTH"'), default="")
    log("AzuraCast Live Disconnected Response: #{ret}")

    live_enabled := false
    last_authenticated_dj := ""
    live_dj := ""
end

# A Pre-DJ source of radio that can be broadcast if needed',
radio_without_live = radio
ignore(radio_without_live)

# Live Broadcasting
live = audio_to_stereo(input.harbor("/", id = "input_streamer", port = 8005, auth = dj_auth, icy = true, icy_metadata_charset = "UTF-8", metadata_charset = "UTF-8", on_connect = live_connected, on_disconnect = live_disconnected, buffer = 15.00, max = 20.00))
ignore(output.dummy(live, fallible=true))

radio = fallback(id="live_fallback", replay_metadata=false, track_sensitive=false, [live, radio])

# Allow for Telnet-driven insertion of custom metadata.
radio = server.insert_metadata(id="custom_metadata", radio)

# Apply amplification metadata (if supplied)
radio = amplify(override="liq_amplify", 1., radio)

radio = fallback(id="safe_fallback", track_sensitive = false, [radio, single(id="error_jingle", "/usr/local/share/icecast/web/error.mp3")])

# Send metadata changes back to AzuraCast
def metadata_updated(m) =
    def f() =
        if (m["song_id"] != "") then
            ret = list.hd(get_process_lines(env=[("SONG", m["song_id"]), ("MEDIA", m["media_id"]), ("PLAYLIST", m["playlist_id"]), ("API_AUTH", !azuracast_api_auth)], 'curl -s --request POST --url http://web/api/internal/1/feedback --form song="$SONG" --form media="$MEDIA" --form playlist="$PLAYLIST" --form api_auth="$API_AUTH"'), default="")
            log("AzuraCast Feedback Response: #{ret}")
        end
        (-1.)
    end

    add_timeout(fast=false, 0., f)
end

radio = on_metadata(metadata_updated,radio)

radioMONO = mean(radio)
radioOPUS = drop_metadata(radio)
radioOPUSmono = mean(radioOPUS)
radioMONOaac = audio_to_stereo(radioMONO)

def remove_unknown(radio)
  # Remove (_,"Unknown") metadata
  def f(m) =
    def unknown_meta(el) =
      let (_, m) = el
      m != "Unknown"
    end

    list.filter(unknown_meta, m)
  end

  map_metadata(update=false, f, radio)
end

radio = remove_unknown(radio)

# Local Broadcasts
output.icecast(%mp3(samplerate=44100, stereo=true, bitrate=128, id3v2=true), id="local_1", host = "127.0.0.1", port = 8000, password = "(PASSWORD)", mount = "/1", name = "Radio Paranormalium", description = "", genre = "", public = false, encoding = "UTF-8", radio)
output.icecast(%fdkaac(channels=2, samplerate=44100, bitrate=320, afterburner=true, aot="mpeg4_aac_lc", sbr_mode=false), id="local_2", host = "127.0.0.1", port = 8000, password = "(PASSWORD)", mount = "/2", name = "Radio Paranormalium", description = "", genre = "", public = true, encoding = "UTF-8", radio)
output.icecast(%opus(samplerate=48000, bitrate=192, vbr="none", application="audio", channels=2, signal="music", complexity=10, max_bandwidth="full_band"), id="local_3", host = "127.0.0.1", port = 8000, password = "(PASSWORD)", mount = "/opusHQ", name = "Radio Paranormalium", description = "", genre = "", public = true, encoding = "UTF-8", icy_metadata="true", radioOPUS)
output.icecast(%mp3(samplerate=22050, stereo=false, bitrate=48, id3v2=true), id="local_4", host = "127.0.0.1", port = 8000, password = "(PASSWORD)", mount = "/mp3_48", name = "Radio Paranormalium", description = "", genre = "", public = true, encoding = "UTF-8", radioMONO)
output.icecast(%fdkaac(channels=2, samplerate=44100, bitrate=24, afterburner=false, aot="mpeg4_he_aac_v2", sbr_mode=true), id="local_5", host = "127.0.0.1", port = 8000, password = "(PASSWORD)", mount = "/5", name = "Radio Paranormalium", description = "", genre = "", public = true, encoding = "UTF-8", radioMONOaac)
output.icecast(%opus(samplerate=48000, bitrate=24, vbr="none", application="audio", channels=1, signal="music", complexity=10, max_bandwidth="full_band"), id="local_6", host = "127.0.0.1", port = 8000, password = "(PASSWORD)", mount = "/opus24", name = "Radio Paranormalium", description = "", genre = "", public = true, encoding = "UTF-8", icy_metadata="true", radioOPUSmono)
output.icecast(%mp3(samplerate=22050, stereo=false, bitrate=24, id3v2=true), id="local_7", host = "127.0.0.1", port = 8000, password = "(PASSWORD)", mount = "/6", name = "Radio Paranormalium", description = "", genre = "", public = true, encoding = "UTF-8", radioMONO)
output.icecast(%opus(samplerate=48000, bitrate=8, vbr="none", application="audio", channels=1, signal="music", complexity=10, max_bandwidth="full_band"), id="local_8", host = "127.0.0.1", port = 8000, password = "(PASSWORD)", mount = "/opus8", name = "Radio Paranormalium", description = "", genre = "", public = true, encoding = "UTF-8", icy_metadata="true", radioOPUSmono)

# Remote Relays
`

**Comment by ivellios1988** [comment]
Here's my Liquidsoap configuration, maybe this will help finding the culprit:

https://pastebin.com/h3JWc4Ak

**Comment by ivellios1988** [comment]
Well, it looks like I accidentally found the culprit...

My radio's list of streams looks as follows:
MP3 128kbps
AAC 320kbps
OPUS 192kbps VBR
MP3 48kbps
AAC+ 24kbps
MP3 24kbps
OPUS 24kbps
OPUS 8kbps

I know there are plenty of them. I wanted to fit the needs of all my listeners. Some of them (most of my listeners live in Poland) have a bad Internet connection so they will use streams with 24kbps or 8kbps, others prefer the highest audio fidelity so they use OPUS 192kbps or AAC 320kbps. Those who still sit on Windows XP or Vista will probably choose MP3 streams.... etc.

Now, about OPUS streams. For a long time, there was an issue with the playback of OPUS streams in web players. On each metadata update, the streams were interrupted which resulted in browsers stopping the playback. I wanted to prevent this from happening so I decided to add drop_metadata to OPUS streams so that the playback won't be interrupted.

Now it turns out that probably this is what caused this 'unknown' metadata issue with some of my tracks. After I updated my installation of AzuraCast to 0.15.0 two days ago, Liquidsoap started its' work from setting metadata to 'unknown' (very funny...). Because I don't want my listeners to be annoyed by this anymore, I decided to remove the 'drop_metadata' argument. And the issue is gone, yaaaaay! I keep monitoring the title bar in my player and the song history - no 'unknowns'.

And apparently, something has been changed in how Chrome deals with OPUS streams (i haven't tested playing them in Firefox yet), because metadata updates don't cause playback interruptions anymore. So another good news is that I don't need to use drop_metadata anymore.
